docs: add insurance arbitrator (#9)
UMAprotocol/uma-docs: developers/in-depth-tutorial-insurance-claims-arbitration.md

# In depth tutorial: Insurance claims arbitration

This section covers [Insurance claims arbitration contract](https://github.com/UMAprotocol/dev-quickstart/blob/main/contracts/InsuranceArbitrator.sol), which is available at the [developer's quick-start repo](https://github.com/UMAprotocol/dev-quickstart). This tutorial shows example on how insurance claims can be resolved and settled through [Optimistic Oracle V2](https://github.com/UMAprotocol/protocol/blob/master/packages/core/contracts/oracle/implementation/OptimisticOracleV2.sol).

You will find out how to test and deploy this smart contract and how it integrates with Optimistic Oracle. Please see more details on [how does UMA's Oracle work](../protocol-overview/how-does-umas-oracle-work.md) section.

### Insurance Arbitrator Contract

This smart contract allows insurers to issue insurance policies by depositing insured amount, designating the insured beneficiary and describing insured event.

> At any time anyone can submit a claim that the insured event has occurred by posting oracle bonding. Insurance Arbitrator resolves the claim through Optimistic Oracle by passing a templated question with insured event description in ancillary data using `YES_OR_NO_QUERY` price identifier as specified in [UMIP-107](https://github.com/UMAprotocol/UMIPs/blob/master/UMIPs/umip-107.md).

Reviewer: ```suggestion
Anyone can submit a claim that the insured event has occurred at any time. Insurance Arbitrators resolves the claim through Optimistic Oracle by passing a templated question with the insured event description in ancillary data using `YES_OR_NO_QUERY` price identifier as specified in [UMIP-107](https://github.com/UMAprotocol/UMIPs/blob/master/UMIPs/umip-107.md).
```

> If the claim is confirmed and settled through Optimistic Oracle this contract automatically pays out insurance coverage to the insured beneficiary. If the claim is rejected the policy continues to be active ready for the subsequent claim attempts.

Reviewer: ```suggestion
If the claim is confirmed and settled through Optimistic Oracle this contract automatically pays out insurance coverage to the insured beneficiary. If the claim is rejected the policy continues to be active and ready for subsequent claim attempts.
```

> ### Development environment and tests

Reviewer: we should show how to do this via remix as well! I can do this via a subsequent PR

#### Install dependencies

> To install dependencies, you will need to install the long-term support version of nodejs, currently nodejs v16, and yarn. You can then install dependencies by running yarn with no arguments from the root of [developer's quick-start repo](https://github.com/UMAprotocol/dev-quickstart):

Reviewer: ```suggestion
To install dependencies, you will need to install the long-term support version of Nodejs, currently Nodejs v16, and yarn. You can then install dependencies by running yarn with no arguments from the root of [developer's quick-start repo](https://github.com/UMAprotocol/dev-quickstart):
```

```bash
yarn
```

#### Compiling your contracts

> We will need to run the following command to compile the contracts and make the typescript interfaces so that they are easy to work with:

Reviewer: ```suggestion
We will need to run the following command to compile the contracts and make the Typescript interfaces so that they are easy to work with:
```

```bash
yarn hardhat compile
```

### Contract implementation

#### Contract creation and initialization

> `_finderAddress` parameter in the constructor points the Insurance Arbitrator to the entry point of the rest of UMA environment. This can be either fetched from the relevant [networks](https://github.com/UMAprotocol/protocol/tree/master/packages/core/networks) file looking up the address of deployed `Finder` contract or you can provide your own `Finder` instance if deploying UMA [protocol](https://github.com/UMAprotocol/protocol) in your own sandboxed environment.

Reviewer: ```suggestion
`_finder` parameter in the constructor points the Insurance Arbitrator to the entry point of the rest of UMA environment. This can be either fetched from the relevant [networks](https://github.com/UMAprotocol/protocol/tree/master/packages/core/networks) file looking up the address of deployed `Finder` contract or you can provide your own `Finder` instance if deploying UMA [protocol](https://github.com/UMAprotocol/protocol) in your own sandboxed environment.
```

`_currency` parameter in the constructor allows choosing the settlement and bonding currency for insurance claims. This should be approved as whitelisted UMA collateral. Please check [Approved Collateral Types](../resources/approved-collateral-types.md) for production networks or call `getWhitelist()` for any of test networks to the [Address Whitelist](https://github.com/UMAprotocol/protocol/blob/master/packages/core/contracts/common/implementation/AddressWhitelist.sol) contract. Alternatively, one can approve the token with `addToWhitelist` method to the Address Whitelist contract if working in a sandboxed UMA environment.

> `_timerAddress` is used only when running unit tests locally to simulate advancement of time. For all the public networks (including testnets) zero address should be used.

Reviewer: ```suggestion
`_timer` is used only when running unit tests locally to simulate the advancement of time. For all the public networks (including testnets) zero address should be used.
```

Reviewer: the three changes below are to match with the implementation changes made in this PR: https://github.com/UMAprotocol/dev-quickstart/pull/18 that I did to clean this up a little bit.

```solidity
    constructor(

>         address _finderAddress,

Reviewer: ```suggestion
        FinderInterface _finder,
```

        address _currency,

>         address _timerAddress

Reviewer: ```suggestion
        address _timer
```

    ) Testable(_timerAddress) {

>         finder = FinderInterface(_finderAddress);

Reviewer: ```suggestion
        finder =_finderAddress;
```

        oo = OptimisticOracleV2Interface(finder.getImplementationAddress(OracleInterfaces.OptimisticOracleV2));
    }
```

As part of initialization `oo` variable is set to the address of `OptimisticOracleV2` implementation as discovered through `getImplementationAddress` method in the [Finder](https://github.com/UMAprotocol/protocol/blob/master/packages/core/contracts/oracle/implementation/Finder.sol) contract.

#### Issuing insurance

`issueInsurance` method allows any insurer to deposit `insuredAmount` of `currency` tokens by designating insurance beneficiary (`insuredAddress`) and defining the insured event (`insuredEvent`). Before calling this method the insurer should have approved this contract to spend required amount of `currency` tokens.

```solidity
    function issueInsurance(
        string calldata insuredEvent,
        address insuredAddress,
        uint256 insuredAmount
    ) external returns (bytes32 policyId) ...
```

Internally the issued policy is stored in the `insurancePolicies` mapping using the calculated `policyId` key that is generated by hashing current block number with provided insurance parameters in internal `_getPolicyId` function.

> After pulling `insuredAmount` from the caller the `issueInsurance` method emits `PolicyIssued` event including `policyId` parameter that should be used when claiming insurance.

Reviewer: ```suggestion
After pulling `insuredAmount` from the caller the `issueInsurance` method, emit `PolicyIssued` event including `policyId` parameter that should be used when claiming insurance.
```

#### Submitting insurance claim

Anyone can submit insurance claim on the issued policy by calling `submitClaim` method with the relevant `policyId` parameter. This method will initiate both data request and proposal at the Optimistic Oracle that requires bonding, hence the caller should have approved this contract to spend required amount of `currency` tokens.

```solidity
    function submitClaim(bytes32 policyId) ...
```

After checking that the `policyId` represents a valid unclaimed insurance policy the contract gets the current `timestamp` and composes `ancillaryData` that will be required for making request and proposal on Optimistic Oracle:

```solidity

>         uint256 timestamp = getCurrentTime();

Reviewer: ```suggestion
        uint256 timestamp = getCurrentTime(); // note that `getCurrentTime` is exported from testable to enable easy time manipulation.
```

        bytes memory ancillaryData = abi.encodePacked(ancillaryDataHead, claimedPolicy.insuredEvent, ancillaryDataTail);
        bytes32 claimId = _getClaimId(timestamp, ancillaryData);
        insuranceClaims[claimId] = policyId;
```

The resulting `timestamp` and `ancillaryData` parameters are hashed in the internal `_getClaimId` method that is used as a key when storing the linked `policyId` in the `insuranceClaims` mapping. This information will be required when receiving callback from Optimistic Oracle.

As a result the concatenated `ancillaryData` will have a valid question as specified in [UMIP-107](https://github.com/UMAprotocol/UMIPs/blob/master/UMIPs/umip-107.md) for `YES_OR_NO_QUERY` price identifier.

Optimistic Oracle data request is initiated with above calculated parameters without providing any proposer reward since the proposal will be done in this `submitClaim` method:

```solidity
        oo.requestPrice(priceIdentifier, timestamp, ancillaryData, currency, 0);
```

> Before the proposal is made Optimistic Oracle allows the requesting contract (this Insurance Arbitrator) to set additional parameters like bonding, liveness and callback settings. This requires passing the same `priceIdentifier`, `timestamp` and `ancillaryData` parameters to identify the request.

Reviewer: ```suggestion
Before the proposal is made the Optimistic Oracle allows the requesting contract (this Insurance Arbitrator) to set additional parameters like bonding, liveness, and callback settings. This requires passing the same `priceIdentifier`, `timestamp`, and `ancillaryData` parameters to identify the request.
```

Total bond to be pulled from the claim initiator consists of Optimistic Oracle proposer bond and final fee for the relevant `currency` token. This contract sets proposer bond as fixed percentage (constant `oracleBondPercentage`) from `insuredAmount`. When calling `setBond` method the Optimistic Oracle calculates and returns the total bond that would be pulled when making proposal:

```solidity
        uint256 proposerBond = (claimedPolicy.insuredAmount * oracleBondPercentage) / 1e18;
        uint256 totalBond = oo.setBond(priceIdentifier, timestamp, ancillaryData, proposerBond);
```

Optimistic Oracle liveness is set by calling `setCustomLiveness` method. This contract uses 24h so that verifiers have sufficient time to check the claim, but one can adjust the `optimisticOracleLivenessTime` constant if required for testing:

```solidity
        oo.setCustomLiveness(priceIdentifier, timestamp, ancillaryData, optimisticOracleLivenessTime);
```

> In contrast to earlier versions Optimistic Oracle V2 by default does not use callbacks and requesting contracts have to explicitly subscribe to them if intending to perform any logic when data request has changed state. Here in calling `setCallbacks` this contract only subscribes to callback only for when the request will have settled at the Optimistic Oracle as implemented in the `priceSettled` method. Subscribing to any other callbacks that are not implemented at the requesting contract would make data requests unresolvable.

Reviewer: ```suggestion
In contrast to earlier versions, the Optimistic Oracle V2 by default does not use callbacks and requesting contracts have to explicitly subscribe to them if intending to perform any logic when data request has changed state. Here, in calling `setCallbacks`, this contract only subscribes to callback only for when the request will have settled at the Optimistic Oracle as implemented in the `priceSettled` method. Subscribing to any other callbacks that are not implemented at the requesting contract would make data requests unresolvable.
```

```solidity
        oo.setCallbacks(priceIdentifier, timestamp, ancillaryData, false, false, true);
```

> After `totalBond` amount of `currency` token is pulled from the claim initiator and approved to be taken by Optimistic Oracle this contract proposes `1e18` representing the `YES` answer to the raised question. Requesting and proposing affirmative answer atomically allows to reduce number of steps to be taken by end users and it is most likely expected that the insured beneficiary would be initiating the claim.

Reviewer: ```suggestion
After `totalBond` amount of `currency` token is pulled from the claim initiator and approved to be taken by Optimistic Oracle, this contract proposes `1e18` representing the `YES` answer to the raised question. Requesting and proposing affirmative answers atomically allows us to reduce the number of steps taken by end users and it is most likely expected that the insured beneficiary would be initiating the claim.
```

```solidity
        oo.proposePriceFor(msg.sender, address(this), priceIdentifier, timestamp, ancillaryData, int256(1e18));
```

#### Disputing insurance claim

For the sake of simplicity this contract does not implement a dispute method, but the disputer can dispute the submitted claim directly through Optimistic Oracle before the liveness passes by calling its `disputePrice` method:

```solidity
    function disputePrice(
        address requester,
        bytes32 identifier,
        uint256 timestamp,
        bytes memory ancillaryData
    ) ...
```

Disputer should pass the address of this Insurance Arbitrator contract as `requester` and all the other parameters from the original request when claim was initiated as emitted by Optimistic Oracle in its `RequestPrice` event.

> If the claim is disputed the request is escalated to UMA DVM and it can be settled only after UMA voters have resolved it.

Reviewer: ```suggestion
If the claim is disputed then therequest is escalated to UMA DVM and it can be settled only after UMA voters have resolved it.
```

#### Settling insurance claim

> Similarly as dispute the claim settlement should be initiated through Optimistic Oracle by calling its `settle` method with the same parameters:

Reviewer: ```suggestion
Similarly, as dispute the claim settlement should be initiated through Optimistic Oracle by calling its `settle` method with the same parameters:
```
